Fix HTTPS upload

frauzufall opened the pull request:
This PR should fix the upload via HTTPS.

It removes the current implementation using a reflection hack to enable WEBDAV methods such as `PROPFIND`. This only worked with the `HttpURLConnection`, not with `HttpsURLConnection`. 

The new implementation is based on Apache's library `jackrabbit-webdav`. 

References:
Bug report for Java 8: https://bugs.openjdk.java.net/browse/JDK-7016595
Should be fixed in new API (Java 11): https://dzone.com/articles/java-11-standardized-http-client-api

frauzufall commented:
@ctrueden I added a dependency to a SNAPSHOT version of `jackrabbit-webdav`. Seems that [`3.0-SNAPSHOT` is from 2014](https://repository.apache.org/content/groups/snapshots/org/apache/jackrabbit/jackrabbit-webdav/3.0-SNAPSHOT/jackrabbit-webdav-3.0-20140729.170246-8) though. I wanted to change it to a release version, but the older 2.X versions work different and I like the current implementation. Can we keep it or should I investigate further?

ctrueden commented:
> Can we keep it or should I investigate further?

Two options:

1. Find out why jackrabbit-webdav was discontinued. Is there a replacement? Is there another wagon that works better these days? If so, test it out, and hopefully switch to it.
2. Assuming jackrabbit-webdav 3.0-SNAPSHOT is really what we continue to prefer, we can deploy our own release version (fork) to our Maven repository and use that.

imagejan commented:
The latest 2.x version of `jackrabbit-webdav` is 2.19.1 and it seems to be from Feb 2019:
https://mvnrepository.com/artifact/org.apache.jackrabbit/jackrabbit-webdav
Why do you call it "old"?

And where did you find a 3.0.0-SNAPSHOT, @frauzufall? The link you posted gives a `404` error for me.

imagejan commented on pom.xml:
```diff
@@ -5,13 +5,13 @@
 	<groupId>net.imagej</groupId>
 	<artifactId>imagej-plugins-uploader-webdav</artifactId>
-	<version>0.2.3-SNAPSHOT</version>
+	<version>0.3.0-SNAPSHOT</version>
```
Can we define properties for these and then use them as `<version>${imagej-plugins-uploader-webdav.version}</version>`, to allow them to be overridden by tooling such as the melting pot script?

ctrueden replied:
For the parent and artifact `<version>` values we do not (and should not) use properties. But for dependency version declarations, we should!

imagejan commented on pom.xml:
```diff
@@ -98,6 +98,7 @@ Institute of Molecular Cell Biology and Genetics.</license.copyrightOwners>
 		<dependency>
 			<groupId>net.imagej</groupId>
 			<artifactId>imagej-updater</artifactId>
+			<version>0.10.0-SNAPSHOT</version>
```
Use version `<property>`, see comment above?

imagejan commented on pom.xml:
```diff
@@ -124,6 +131,7 @@ Institute of Molecular Cell Biology and Genetics.</license.copyrightOwners>
 			<artifactId>imagej-updater</artifactId>
 			<classifier>tests</classifier>
 			<scope>test</scope>
+			<version>0.10.0-SNAPSHOT</version>
```
Use version `<property>`, see comment above?

frauzufall commented:
When we searched for the best way to use webdav from Java, apache jackrabbit came up a lot, and @imagejan you are right, it's updated frequently. I think I searched for the latest version on maven.imagej.net and somehow only saw snapshots so I picked what I thought would be the newest (corrected the link in the text above). That was a bad decision. I know now that the maven mirror only gets populated by what people request.. The 3.0er version is super convenient to use and matched examples I found, that's why I did not recognize the issue earlier. But I will have a closer look at how to migrate to 2.19.1.

frauzufall commented:
Found equivalent release version `2.12.10` for `jackrabbit-webdav` that is not a SNAPSHOT. Will look into how to update this to the newest version, now that I can follow the release changes. @imagejan I added the properties, thanks for the comment, makes a lot of sense. Trying out the melting pot is on my TODO list..

frauzufall commented:
Updated the code to the newest `jackrabbit-webdav` version which meant rewriting the whole thing to be compatible with Apache HttpClient 4. I also updated http://sites.imagej.net/UpdaterBeta/ where one can test it. Now I have no further points on my agenda for this PR.

@axtimwalde I read your last commit about merging the branches on this repo and on `juglab` and thought I could push to `imagej/fix-https-upload`, but that created a new branch. So I also pushed to `juglab/fix-https-upload` to make the changes appear on this page. Sorry for the mess.. Should I delete the redundant `imagej/fix-https-upload` branch again?

ctrueden commented on pom.xml:
```diff
@@ -50,6 +50,12 @@
+		<contributor>
+			<name>Deborah Schmidt</name>
+			<url>http://imagej.net/User:Frauzufall</url>
+			<roles><role>maintainer</role></roles>
```
If you are a `maintainer`, then you are not a `<contributor>` but rather a `<developer>`.

frauzufall replied:
I would like to maintain this, yes. Do you agree or should I rather change my role?

ctrueden replied:
Wonderful—I wholeheartedly support additional maintainers on whichever components you feel are relevant to your work. Please see the [SciJava team roles](https://imagej.net/Team) and add yourself as a `<developer>` which whichever of those roles are appropriate.

ctrueden commented on pom.xml:
```diff
@@ -84,6 +90,13 @@ Institute of Molecular Cell Biology and Genetics.</license.copyrightOwners>
+
+		<!-- skip reproducible build enforcer during development -->
+		<imagej-updater.version>0.10.0-SNAPSHOT</imagej-updater.version>
+		<enforcer.skip>true</enforcer.skip>
```
The commits built on this snapshot version will need to be rewritten before this PR can be merged.

ctrueden commented on src/main/java/net/imagej/plugins/uploaders/webdav/ProgressHttpEntityWrapper.java:
```diff
@@ -0,0 +1,95 @@
+/*
+ * The MIT License (MIT)
```
This file should have our usual license on top, and then the original code's MIT license beneath it. Annoying I know, but that's third party attribution for you...

ctrueden commented on src/test/java/net/imagej/plugins/uploaders/webdav/WebDAVUploaderTests.java:
```diff
@@ -0,0 +1,142 @@
+
+/**
+ * A conditional JUnit test for uploading via WebDAV.
+ * 
```
Need `<p>` tags around subsequent paragraphs. Otherwise, the javadoc does not render as expected. And IDE javadoc cleanup functions will mush the lines together.

But thank you very much for writing very nice, detailed javadoc. 😄

ctrueden reviewed (changes requested):
Thanks very much for working on this! I made some comments. Minor things, except for the SNAPSHOT dependency which will need to be resolved in the history before we can merge.

frauzufall committed:
WebDAVUploader: Integration of Apache HTTP / jackrabbit-webdav library

* sets imagej-updater dependency to version 0.10.1
* adds dependency to org.apache.jackrabbit:jackrabbit-webdav 2.19.1
* adds dependency to org.apache.httpcomponents:httpclient 4.5.6
* adds dependency to org.apache.httpcomponents:httpcore 4.4.11
* fixes HTTPS upload
* adds implementation for test checking if file is deleted
* made a couple of WebDAVUploader method package private for testing
* adds more tests (which still require url and credentials system
properties, therefore no automated Travis testing for now)
* tested with digest and basic authentication

frauzufall committed:
POM: add developer
